src/containers/auth/AuthProvider.js

```diff
@@ -0,0 +1,178 @@
+import {
+  createContext,
+  useState,
+  useContext,
+  useEffect,
+  useCallback,
+  useMemo,
+} from "react";
+import { useSelector, useDispatch } from "react-redux";
+import { useNavigate, useLocation } from "react-router-dom";
+import PropTypes from "prop-types";
+
+import { jwtDecode } from "jwt-decode";
+
+import * as authAPIs from "../../apis/auth";
+import { setAccessToken, setRoleToken, hasAuthTokens } from "../../utils/auth";
+
+import * as authSelectors from "../../store/selectores/auth";
+import { showSuccessAlert } from "../../store/directDispatches/common";
+import * as userActions from '../../store/actions/user';
+
+const AuthContext = createContext({});
+
+function AuthProvider(props) {
+  const { children } = props;
+
+  /* -------------------------------------------------------------------------- */
+  /*                                    Redux                                   */
+  /* -------------------------------------------------------------------------- */
+  const dispatch = useDispatch();
+
+  const [auth, setAuth] = useState({});
+  const [userPermissions, setUserPermissions] = useState([]);
+  const [isBackendAuthorized, setIsBackendAuthorized] = useState(
+    hasAuthTokens()
+  );
+  const [isProcessing, setIsProcessing] = useState(false);
+  const location = useLocation();
+  const navigate = useNavigate();
+  const isTokenExpired = useSelector(authSelectors.isTokenExpired);
+
+  const loginUserFlow = useCallback(
+    async (user, pwd) => {
+      try {
+        const from = location.state?.from?.pathname || "/";
+        setIsProcessing(true);
+        const response = await authAPIs.loginUser({
+          email: user,
+          password: pwd,
+        });
+        console.log("token", response?.data?.token);
+        console.log("roleToken", response?.data?.roleToken);
+        const accessToken = response?.data?.token;
+        const roleToken = response?.data?.roleToken;
+        const decoded = roleToken ? jwtDecode(roleToken) : undefined;
+        if (decoded !== undefined) {
+          const roles = decoded?.roles || [];
+          console.log("decoded", roles);
+          setUserPermissions(roles);
+          setAuth({ user, accessToken });
+          setAccessToken(accessToken);
+          setRoleToken(roleToken);
+          dispatch(
+            userActions.setUserData({
+              id: response?.data?.user?.id,
+              name: response?.data?.user?.email,
+            }),
+          );
+          setIsBackendAuthorized(true);
+          setIsProcessing(false);
+          navigate(from, { replace: true }); // navigate to the page where user wanted to go before navigationg to login.
+        } else {
+          setIsBackendAuthorized(false);
+          setUserPermissions([]);
+          setIsProcessing(false);
+          navigate("/login");
+        }
+      } catch (err) {
+        setIsProcessing(false);
+      }
+    },
+    [location.state?.from?.pathname, navigate, dispatch]
```

Comment:
do we need to add navigate, dispatch as dependancies?

Reply:
navigate and dispatch are added because of eslint missing dependency warning. But these are not changing as per every cycle and we will be able to remove them.

Reply:
Fixed